tsdb/index/tsi1/index.go

```diff
@@ -618,81 +520,164 @@ func (i *Index) MeasurementsSketches() (estimator.Sketch, estimator.Sketch, erro
+		if err := <-errC; err != nil {
+			return false, err
+		}
+	}
+
+	// Check if we found the tag key.
+	return atomic.LoadUint32(&found) == 1, nil
 }
 
 // MeasurementTagKeysByExpr extracts the tag keys wanted by the expression.
 func (i *Index) MeasurementTagKeysByExpr(name []byte, expr influxql.Expr) (map[string]struct{}, error) {
-	fs := i.RetainFileSet()
-	defer fs.Release()
-	return fs.MeasurementTagKeysByExpr(name, expr)
+	n := i.availableThreads()
+
+	// Store results.
+	keys := make([]map[string]struct{}, i.PartitionN)
+	errC := make(chan error, i.PartitionN)
+
+	var pidx uint32 // Index of maximum Partition being worked on.
+	var err error
+	for k := 0; k < n; k++ {
+		go func() {
+			for {
+				idx := int(atomic.AddUint32(&pidx, 1) - 1) // Get next partition to work on.
+				if idx >= len(i.partitions) {
+					return // No more work.
+				}
+
+				// This is safe since there are no readers on keys until all
+				// the writers are done.
+				keys[idx], err = i.partitions[idx].MeasurementTagKeysByExpr(name, expr)
+				errC <- err
+			}
+		}()
+	}
+
+	// Check for error
+	for i := 0; i < cap(errC); i++ {
+		if err := <-errC; err != nil {
+			return nil, err
+		}
+	}
+
+	// Merge into single map.
+	result := keys[0]
+	for k := 1; k < len(i.partitions); k++ {
+		for k := range keys[k] {
+			result[k] = struct{}{}
+		}
+	}
+	return result, nil
 }
 
 // MeasurementTagKeyValuesByExpr returns a set of tag values filtered by an expression.
 //
 // See tsm1.Engine.MeasurementTagKeyValuesByExpr for a fuller description of this
 // method.
 func (i *Index) MeasurementTagKeyValuesByExpr(auth query.Authorizer, name []byte, keys []string, expr influxql.Expr, keysSorted bool) ([][]string, error) {
-	fs := i.RetainFileSet()
-	defer fs.Release()
-
 	if len(keys) == 0 {
 		return nil, nil
 	}
 
-	results := make([][]string, len(keys))
 	// If we haven't been provided sorted keys, then we need to sort them.
 	if !keysSorted {
 		sort.Sort(sort.StringSlice(keys))
 	}
 
+	resultSet := make([]map[string]struct{}, len(keys))
+	for i := 0; i < len(resultSet); i++ {
+		resultSet[i] = make(map[string]struct{})
+	}
+
 	// No expression means that the values shouldn't be filtered, so we can
 	// fetch them all.
-	if expr == nil {
-		for ki, key := range keys {
-			itr := fs.TagValueIterator(name, []byte(key))
-			if auth != nil {
-				for val := itr.Next(); val != nil; val = itr.Next() {
-					si := fs.TagValueSeriesIterator(name, []byte(key), val.Value())
-					for se := si.Next(); se != nil; se = si.Next() {
-						if auth.AuthorizeSeriesRead(i.Database, se.Name(), se.Tags()) {
-							results[ki] = append(results[ki], string(val.Value()))
-							break
+	for _, p := range i.partitions {
+		if err := func() error {
+			fs := p.RetainFileSet()
+			defer fs.Release()
+
+			if expr == nil {
+				for ki, key := range keys {
+					itr := fs.TagValueIterator(name, []byte(key))
+					if itr == nil {
+						continue
+					}
+					if auth != nil {
```

Comment:
Rather than having everything indented, how about just indenting the smallest block?

```go
if auth == nil {
        for val := itr.Next(); val != nil; val = itr.Next() {
                resultSet[ki][string(val.Value())] = struct{}{}
        }
        continue
}

// Auth base stuff?
```